File: backend/src/main/java/com/training/cvmanagementbe/service/CvService.java
package com.training.cvmanagementbe.service;

import com.training.cvmanagementbe.dto.request.CvCreateRequest;
import com.training.cvmanagementbe.dto.request.CvDeleteRequest;
import com.training.cvmanagementbe.dto.request.CvEditRequest;
import com.training.cvmanagementbe.dto.response.*;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

public interface CvService {

    /*
     * Creates a CV in one profile and language, snapshotting personal_info from the user record
     * and auto-linking the oldest matching PENDING update request.
     */
    CvResponse create(UUID profileId, CvCreateRequest request);

    // Returns the current version - the row with the highest version_number.
    CvDetailResponse getById(UUID cvId);

    /*
     * Writes CV content. Only the owner may call it; the owner's role decides whether
     * the write lands in a draft or becomes a version straight away.
     */
    CvEditResponse edit(UUID cvId, CvEditRequest request);

    // Soft-deletes the CV. Blocked while a draft awaits approval; a master needs a successor.
    void delete(UUID cvId, CvDeleteRequest request);

    /*
     * Restores a soft-deleted CV once the three conditions hold:
     * 1. The (profile, language) slot is not already occupied by another ACTIVE CV (CV_SLOT_OCCUPIED).
     * 2. The parent profile is currently ACTIVE; if deleted, the profile must be restored first (CV_PROFILE_DELETED).
     * 3. Valid Master relationship: If the CV was previously a Master,
     *    its master_cv_id must be updated to point to the profile's current active Master within
     *    the same transaction (to avoid having two active masters, CV_MASTER_CONFLICT).
     *    If the profile currently has no other active CVs, keep master_cv_id null and it becomes Master.
     */
    CvResponse restore(UUID cvId);

    List<CvResponse> listByProfile(UUID profileId, boolean includeDeleted);

    PagedResponse<CvResponse> listDeleted(Pageable pageable);

    List<CvVersionSummary> listVersions(UUID cvId);
}

File: frontend/src/app/core/services/cv.service.ts
import { HttpClient, HttpParams } from "@angular/common/http";
import { inject, Injectable } from "@angular/core";
import { environment } from "../../../environments/environment";
import { Observable } from "rxjs";
import { CvCreateRequest, CvDeleteRequest, CvDetailResponse, CvEditRequest, CvEditResponse, CvResponse, CvVersionSummary, DeletedCvQuery } from "../dtos/cv.dto";
import { ApiEndpoint } from "../enums/api-endpoint.enum";
import { PagedResponse } from "../dtos/page.dto";

@Injectable({ providedIn: 'root' })
export class CvService {

    private readonly http = inject(HttpClient);

    private url(endpoint: string): string {
        return `${environment.apiBaseUrl}${endpoint}`;
    }

    listByProfile(profileId: string, includeDeleted = false): Observable<CvResponse[]> {
        const params = new HttpParams().set('includeDeleted', includeDeleted);
        return this.http.get<CvResponse[]>(this.url(`${ApiEndpoint.Profiles}/${profileId}/cvs`), { params });
    }

    listDeleted(query: DeletedCvQuery): Observable<PagedResponse<CvResponse>> {
        let params = new HttpParams()
            .set('page', query.page)
            .set('size', query.size);

        if (query.sortBy) {
            params = params.set('sortBy', query.sortBy);
        }
        if (query.direction) {
            params = params.set('direction', query.direction);
        }
        return this.http.get<PagedResponse<CvResponse>>(
            this.url(`${ApiEndpoint.Cvs}/deleted`), { params }
        );
    }
    
    getById(id: string): Observable<CvDetailResponse> {
        return this.http.get<CvDetailResponse>(this.url(`${ApiEndpoint.Cvs}/${id}`));
    }

    listVersions(id: string): Observable<CvVersionSummary[]> {
        return this.http.get<CvVersionSummary[]>(this.url(`${ApiEndpoint.Cvs}/${id}/versions`));
    }

    create(profileId: string, body: CvCreateRequest): Observable<CvResponse> {
        return this.http.post<CvResponse>(this.url(`${ApiEndpoint.Profiles}/${profileId}/cvs`), body);
    }

    edit(id: string, body: CvEditRequest): Observable<CvEditResponse> {
        return this.http.put<CvEditResponse>(this.url(`${ApiEndpoint.Cvs}/${id}/content`), body);
    }

    // A body is sent only when a successor master must be named; the server rejects a missing one.
    delete(id: string, body?: CvDeleteRequest): Observable<void> {
        return this.http.delete<void>(this.url(`${ApiEndpoint.Cvs}/${id}`), { body: body ?? {} });
    }

    restore(id: string): Observable<CvResponse> {
        return this.http.post<CvResponse>(this.url(`${ApiEndpoint.Cvs}/${id}/restore`), null);
    }
}